CreateAndMaintainEphemeral maintains an ephemeral node

burke opened the pull request:
Review: @marc-barry @thegedge 
CC: @Shopify/locutus

burke committed:
CreateAndMaintainEphemeral maintains an ephemeral node

thegedge commented on session/ephemeral.go:
```diff
@@ -0,0 +1,51 @@
+	}
+
+	evs := make(chan ZKSessionEvent)
+	z.Subscribe(ev)
```
s/ev/evs

thegedge commented on session/ephemeral.go:
```diff
@@ -0,0 +1,51 @@
+			// ephemeral node.
+			if err := doCreate(); err != nil {
+				return err
+			}
```
If you'd like to be comprehensive, there's also `SessionFailed`

thegedge commented:
Couple of small things, but otherwise LGTM

burke commented:
I'll see if I can write some reasonable tests for this too.

marc-barry commented on session/ephemeral.go:
```diff
@@ -0,0 +1,57 @@
+package session
```
I don't think this belongs in package `session`. I actually think it belongs in its own package. Perhaps a `util` package since it us general purpose.

marc-barry commented on session/ephemeral.go:
```diff
@@ -0,0 +1,57 @@
+	evs := make(chan ZKSessionEvent)
+	z.Subscribe(evs)
+
+	go func() { dead <- maintainEphemeral(evs, doCreate) }()
```
Do you also want to make this cancellable? Perhaps with an abort channel. I know this will complicate the function signature but as it currently stands there is no way for the caller to abort maintaining the ephemeral node.

marc-barry commented:
Looks pretty good to me. Some minor comments. It seems like a fairly common pattern which this wraps nicely.

burke committed:
Fix implementation bugs in ephemeral

burke committed:
Use system classpath to launch ZK

burke committed:
Apply a timeout to ZK connection establishment

burke committed:
Better test failure behaviour when toxiproxy isn't running

burke committed:
Better handling of indefinite disconnects

burke committed:
Move to new package

burke commented:
I made a bunch of improvements:
- Added `toxiproxy` test to verify this works
- Improved session establishment timeout handling
- Improved handling of extended interruption in ZK connectivity
- Moved to a new package (`ephemeral.CreateAndMaintain(z, "/znode", "")`)

PTAL

marc-barry commented:
LGTM
